helix-term/src/ui/completion.rs

```diff
@@ -82,18 +87,101 @@ impl menu::Item for CompletionItem {
                     ""
                 }
                 None => "",
-            }),
+            },
+            CompletionItem::Path(PathCompletionItem { kind, .. }) => kind.as_str(),
+        };
+
+        menu::Row::new(vec![
+            menu::Cell::from(Span::styled(
+                label,
+                if deprecated {
+                    Style::default().add_modifier(Modifier::CROSSED_OUT)
+                } else {
+                    Style::default()
+                },
+            )),
+            menu::Cell::from(kind),
         ])
     }
 }
 
-#[derive(Debug, PartialEq, Default, Clone)]
-pub struct CompletionItem {
+#[derive(Debug, PartialEq, Clone)]
+pub enum PathKind {
+    Folder,
+    File,
+    Link,
+    Block,
+    Socket,
+    CharacterDevice,
+    Fifo,
+}
+
+impl PathKind {
+    fn as_str(&self) -> &'static str {
+        match self {
+            PathKind::Folder => "folder",
+            PathKind::File => "file",
+            PathKind::Link => "link",
+            PathKind::Block => "block",
+            PathKind::Socket => "socket",
+            PathKind::CharacterDevice => "char_device",
+            PathKind::Fifo => "fifo",
+        }
+    }
+}
+
+impl std::fmt::Display for PathKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.as_str())
+    }
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct LspCompletionItem {
     pub item: lsp::CompletionItem,
     pub provider: LanguageServerId,
     pub resolved: bool,
 }
 
+#[derive(Debug, PartialEq, Clone)]
+pub struct PathCompletionItem {
+    pub kind: PathKind,
+    pub item: helix_core::CompletionItem,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum CompletionItem {
+    Lsp(LspCompletionItem),
+    Path(PathCompletionItem),
```

Comment:
I don't think we need a seperate `PathCompletionItem` type. Instead we can just have a `kind: Cow<str>`  on `helix_core::CompletionItem`. At that point we could also just call that variant `Other` (so it's clear there will only ever be an lsp variant and one for everything else going trough the helix-core type).

I am also not really sure about this abstraction in general. I think first starters this should be much better encapsulated. This file is certainly the wrong place for it (I don't want the handler/resolve code to depend on the UI at all). So probably at least a submodule in `handlers::completion`. I would also like to have as much information extraction as possible within methods (like `preselect`). It's ok to keep `lsp_item_to_transaction in the UI code for now since that is a larger change (altough it also doesn't belong there)

Reply:
> I don't think we need a seperate `PathCompletionItem` type. Instead we can just have a kind: `Cow<str>` on `helix_core::CompletionItem`. At that point we could also just call that variant `Other` (so it's clear there will only ever be an lsp variant and one for everything else going trough the helix-core type).

I just fear that there may be more information relevant for a completion item, but that fear may well be unjustified.

> This file is certainly the wrong place for it

See [recent discussion](https://github.com/helix-editor/helix/pull/2608#discussion_r1720817397) for it, I'm aware of it, but I don't want to start an even bigger refactor, when PRs are open for so long without any kind of review to avoid resolving merge-conflicts. It shouldn't be too hard to move this somewhere else, when that kind of refactor is done.

> At that point we could also just call that variant `Other`.

Reply:
> I just fear that there may be more information relevant for a completion item, but that fear may well be unjustified.

To me if there is more relevant information is should be added to the generic `helix_core::CompletionItem`. I don't want to couple the UI code to the completion sources. Basically `helix_core::CompletionItem` shouldn't be modeled after whatever data completion sources can or can't provide but instead by the data that the UI can actually support (and I would want everything to go trough that generic interface except LSP). The only reason that lsp needs to be separate is because creating the transaction is complicated and needs to be done lazily.

Reply:
> See https://github.com/helix-editor/helix/pull/2608#discussion_r1720817397 for it, I'm aware of it, but I don't want to start an even bigger refactor, when PRs are open for so long without any kind of review to avoid resolving merge-conflicts. It shouldn't be too hard to move this somewhere else, when that kind of refactor is done.

I agree that we shouldn't do a large refactor in this PR but moving this into another file inside helix term (like in the handler) is a pretty easy cosmetic change and reduces spaghetti code by a lot. I don't think that would even conflict with the snippet system PR

Reply:
I think this should be handled.